### PR TITLE
add stale blobs end point

### PR DESCRIPTION
**Pull request by mdayican**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/BPS-1300

### Change description ###

add stale blobs end point

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/blobrouter/controllers/StaleBlobController.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+import java.util.List;
+
+@RestController
+@RequestMapping(path = "/stale-blobs", produces = MediaType.APPLICATION_JSON_VALUE)
+public class StaleBlobController {
+
+    private final BlobLister blobLister;
+
+    private static final String DEFAULT_STALE_TIME = "2";
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/controllers/StaleBlobController.java:19]: what does it mean?

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobLister.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+import static uk.gov.hmcts.reform.blobrouter.util.TimeUtils.toLocalTimeZone;
+
+@Component
+public class BlobLister {
+
+    private final BlobServiceClient storageClient;
+    private final ServiceConfiguration serviceConfiguration;
+
+    public BlobLister(BlobServiceClient storageClient, ServiceConfiguration serviceConfiguration) {
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobLister.java:24]: normally it needs `enable configuration properties`. even though it is enabled somewhere else but instantiation of beans can happen in any order and lazy driven too

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobLister.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    public Predicate<BlobItem> timeFilter(int staleTime) {
+        return blobItem ->
+            Instant
+                .now()
+                .isAfter(
+                    blobItem.getProperties().getCreationTime().toInstant()
+                        .plus(staleTime, ChronoUnit.HOURS)
+                );
+    }
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobLister.java]: perhaps there's no need to to have predicate there. just `staleTime` and everything else can be operated with private methods and method references.

File: src/main/java/uk/gov/hmcts/reform/blobrouter/controllers/StaleBlobController.java (hunk 1 above)
Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/controllers/StaleBlobController.java:19]: request param default value

Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/controllers/StaleBlobController.java:19]: i mean is it hours, days, weeks?

Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/controllers/StaleBlobController.java]: added `hours`

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobLister.java (hunk 3 above)
Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobLister.java]: removed

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobLister.java [hunk 4]
```diff
@@ -26,33 +27,38 @@ public BlobLister(BlobServiceClient storageClient, ServiceConfiguration serviceC
...
-            .map(blob -> new BlobInfo(containerName, blob.getName(),
-                toLocalTimeZone(blob.getProperties().getCreationTime().toInstant())))
+            .filter(b -> timeFilter(b, staleTime))
+            .map(blob -> new BlobInfo(
+                    containerName,
+                    blob.getName(),
+                    toLocalTimeZone(blob.getProperties().getCreationTime().toInstant())
+                )
+            )
             .collect(Collectors.toList());
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobLister.java]: oh, didn't notice before - could just flatmap straight away - just return `Stream` in that first private method and should be all good :+1:

File: src/main/java/uk/gov/hmcts/reform/blobrouter/util/TimeUtils.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import static uk.gov.hmcts.reform.blobrouter.util.TimeZones.EUROPE_LONDON_ZONE_ID;
+
+public final class TimeUtils {
+
+    private TimeUtils() {
+    }
+
+    public static String toLocalTimeZone(Instant instant) {
+        if (instant != null) {
+            DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss");
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/util/TimeUtils.java:16]: can be inlined

File: src/test/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobListerTest.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        given(blobItem.getProperties()).willReturn(properties);
+        if (staleFile) {
+            given(properties.getCreationTime()).willReturn(expectedCreationTime);
+            given(blobItem.getName()).willReturn(name);
+        } else {
+            given(properties.getCreationTime()).willReturn(now());
+        }
+        return blobItem;
+    }
+}
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobListerTest.java:101]: missing new line

File: src/test/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobListerTest.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    private BlobItem blob(String name, boolean staleFile) {
+        var blobItem = mock(BlobItem.class);
+        var properties = mock(BlobItemProperties.class);
+
+        given(blobItem.getProperties()).willReturn(properties);
+        if (staleFile) {
+            given(properties.getCreationTime()).willReturn(expectedCreationTime);
+            given(blobItem.getName()).willReturn(name);
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobListerTest.java:95]: why only when it's stale?

File: src/test/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobListerTest.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+            cmcBlobClient,
+            Stream.of(
+                blob("cmc_scan_file_stale_1", false),
+                blob("cmc_scan_file_stale_2", false)
+            )
+        );
+
+        mockStorageList(sscsBlobClient, Stream.empty());
+
+        List<BlobInfo> blobInfos = blobLister.listBlobs(2);
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobListerTest.java:72]: as commented above, rename class or method name.
`blobLister.listBlobs(2)` does not indicate that this returns stale blobs or blobs that were created more than 2h ago.

File: src/test/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobListerTest.java (hunk 7 above)
Comment (mdayican) [reply src/test/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobListerTest.java]: we get the name if it is stale.

File: src/main/java/uk/gov/hmcts/reform/blobrouter/config/ServiceConfiguration.java [hunk 9]
```diff
@@ -32,4 +44,8 @@ public void setStorageConfig(List<StorageConfigItem> storageConfigItems) {
             .map(StorageConfigItem::getSourceContainer)
             .collect(toList());
     }
+
+    public List<String> getSourceContainers() {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/config/ServiceConfiguration.java:48]: why not use existing `getEnabledSourceContainers`?

Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/config/ServiceConfiguration.java:48]: we may need to get for not enabled ones?

File: src/test/java/uk/gov/hmcts/reform/blobrouter/services/storage/StaleBlobFinderTest.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    @Mock
+    private ServiceConfiguration serviceConfiguration;
+
+    @InjectMocks
+    private StaleBlobFinder staleBlobFinder;
+
+    private OffsetDateTime expectedCreationTime = now().minus(3, ChronoUnit.HOURS);
+
+    @Test
+    void listBlobs() {
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/services/storage/StaleBlobFinderTest.java:43]: please follow existing test naming convention

File: src/main/java/uk/gov/hmcts/reform/blobrouter/config/ServiceConfiguration.java (hunk 9 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/config/ServiceConfiguration.java:48]: Probably not.
If `StaleBlobFinder` is used for alerting, we don't want to be notified about files in disabled containers probably.

File: src/test/java/uk/gov/hmcts/reform/blobrouter/services/storage/StaleBlobFinderTest.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    private OffsetDateTime expectedCreationTime = now().minus(3, ChronoUnit.HOURS);
+
+    @Test
+    void should_find_stale_blobs() {
+        given(serviceConfiguration.getSourceContainers())
+            .willReturn(Arrays.asList("bulkscan", "cmc", "sscs"));
+
+        BlobContainerClient bulkscanBlobClient = mock(BlobContainerClient.class);
+        BlobContainerClient cmcBlobClient = mock(BlobContainerClient.class);
+        BlobContainerClient sscsBlobClient = mock(BlobContainerClient.class);
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/services/storage/StaleBlobFinderTest.java:49]: `var` can be used

File: src/main/java/uk/gov/hmcts/reform/blobrouter/config/ServiceConfiguration.java (hunk 9 above)
Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/config/ServiceConfiguration.java:48]: it is going to be used manually?

File: src/test/java/uk/gov/hmcts/reform/blobrouter/services/storage/StaleBlobFinderTest.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+                blob("bulk_scan_file_new", false),
+                blob("bulk_scan_file_stale", true)
+            )
+        );
+
+        mockStorageList(
+            cmcBlobClient,
+            Stream.of(
+                blob("cmc_scan_file_stale_1", false),
+                blob("cmc_scan_file_stale_2", false)
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/services/storage/StaleBlobFinderTest.java:66]: should those be named stale if they're not stale?

File: src/main/java/uk/gov/hmcts/reform/blobrouter/util/TimeUtils.java [hunk 13]
```diff
@@ -0,0 +1,19 @@
+package uk.gov.hmcts.reform.blobrouter.util;
+
+import java.time.Instant;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+
+import static uk.gov.hmcts.reform.blobrouter.util.TimeZones.EUROPE_LONDON_ZONE_ID;
+
+public final class TimeUtils {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/util/TimeUtils.java:9]: please update `EnvelopeController` once this PR gets merged.

File: src/main/java/uk/gov/hmcts/reform/blobrouter/controllers/StaleBlobController.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    private final StaleBlobFinder staleBlobFinder;
+
+    private static final String DEFAULT_STALE_TIME_HOURS = "2";
+
+    public StaleBlobController(StaleBlobFinder staleBlobFinder) {
+        this.staleBlobFinder = staleBlobFinder;
+    }
+
+    @GetMapping
+    public List<BlobInfo> findStaleBlobs(
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/controllers/StaleBlobController.java:26]: return an object instead of array

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/controllers/StaleBlobController.java:26]: https://opensource.zalando.com/restful-api-guidelines/#110

**Pull request merged**